src/btree/bt_sync.c

```diff
@@ -240,10 +240,15 @@ __sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
 		 * Set the checkpointing flag to block such actions and wait for
 		 * any problematic eviction or page splits to complete.
 		 */
+		WT_ASSERT(session, btree->checkpointing == WT_CKPT_OFF);
+
 		btree->checkpointing = WT_CKPT_PREPARE;
 		(void)__wt_gen_next_drain(session, WT_GEN_EVICT);
 		btree->checkpointing = WT_CKPT_RUNNING;
 
+		/* Note in the session that we are syncing this tree. */
+		session->sync_dhandle = session->dhandle;
+
```

Comment:
This won't get flushed and might get reordered. I don't think it causes problems but it might affect behavior.

Can we move the set above the `__wt_gen_next_drain` call? Pairing it with the initial set of `btree.checkpoint` makes sense to me, it's already paired with the clear.

Reply:
I don't think there's any flushing issue in this version because session handles aren't shared between threads.  That said, I'm about to push a different version that reorders as you suggest.